### PR TITLE
[expo-image] Add ThumbHash support

### PR DESCRIPTION
**Pull request by behenate**
# Why

Previously we were only supporting BlurHash, we want the users to be able to use ThumbHash as well.

# How

Copied the ThumbHash code from the [ThumbHash repo](https://github.com/evanw/thumbhash) and implemented the functionality into `iOS`, `Android` and `web` similarly to BlurHash. Changed the demo app to include more images

# Test Plan

Tested on Android, iOS and Web.

### CODE REVIEW

File: packages/expo-image/android/src/main/java/expo/modules/image/thumbhash/ThumbhashDecoder.kt [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+      while (i < w * h) {
+        val alpha = (rgba[j + 3].toInt() and 255) / 255.0f
+        avg_r += alpha / 255.0f * (rgba[j].toInt() and 255)
+        avg_g += alpha / 255.0f * (rgba[j + 1].toInt() and 255)
+        avg_b += alpha / 255.0f * (rgba[j + 2].toInt() and 255)
+        avg_a += alpha
+        i++
+        j += 4
+      }
+    }
```
Comment (lukmccall) [inline packages/expo-image/android/src/main/java/expo/modules/image/thumbhash/ThumbhashDecoder.kt:37]:
```suggestion
    var i = 0
    var j = 0
    while (i < w * h) {
      val alpha = (rgba[j + 3].toInt() and 255) / 255.0f
      avg_r += alpha / 255.0f * (rgba[j].toInt() and 255)
      avg_g += alpha / 255.0f * (rgba[j + 1].toInt() and 255)
      avg_b += alpha / 255.0f * (rgba[j + 2].toInt() and 255)
      avg_a += alpha
      i++
      j += 4
    }
```
that run block doesn't seem to be needed.

File: packages/expo-image/android/src/main/java/expo/modules/image/thumbhash/ThumbhashDecoder.kt [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+    val lx = Math.max(1, Math.round((l_limit * w).toFloat() / Math.max(w, h).toFloat()))
+    val ly = Math.max(1, Math.round((l_limit * h).toFloat() / Math.max(w, h).toFloat()))
+    val l = FloatArray(w * h) // luminance
+    val p = FloatArray(w * h) // yellow - blue
+    val q = FloatArray(w * h) // red - green
+    val a = FloatArray(w * h) // alpha
+
+    // Convert the image from RGBA to LPQA (composite atop the average color)
+    var i = 0
+    var j = 0
```
Comment (lukmccall) [inline packages/expo-image/android/src/main/java/expo/modules/image/thumbhash/ThumbhashDecoder.kt:54]:
```suggestion
    i = 0
    j = 0
```

File: packages/expo-image/src/utils/resolveHashString.tsx [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import { ImageSource } from '../Image.types';
+
+function hashToUri(type: 'blurhash' | 'thumbhash', hash: string): string {
+  const encodedBlurhash = encodeURI(hash).replace(/#/g, '%23').replace(/\?/g, '%3F');
+  return `${type}:/${encodedBlurhash}`;
```
Comment (tsapeta) [inline packages/expo-image/src/utils/resolveHashString.tsx:5]: Compared to blurhash, is there no way to specify width and height of the image?

**Comment by tsapeta** [comment]
Also, it's a good practice to include a screenshot in the PR description of the examples added to native-component-list, so the reviewers know how it's going to look like 😉

File: packages/expo-image/CHANGELOG.md [hunk 4]
```diff
@@ -21,6 +21,7 @@
 ### 🎉 New features
 
 - [Android] Add automatic asset downscaling to improve performance. ([#21628](https://github.com/expo/expo/pull/21628) by [@lukmccall](https://github.com/lukmccall))
+- Added `ThumbHash` support for iOS, Android and Web. ([#21952](https://github.com/expo/expo/pull/21952) by [@behenate](https://github.com/behenate))
```
Comment (Simek) [inline packages/expo-image/CHANGELOG.md:24]:
```suggestion
- Added `ThumbHash` support for Android, iOS and Web. ([#21952](https://github.com/expo/expo/pull/21952) by [@behenate](https://github.com/behenate))
```

Platform ordering nit.

File: docs/pages/versions/unversioned/sdk/image.mdx [hunk 5]
```diff
@@ -18,7 +18,7 @@ import { YesIcon, NoIcon, PendingIcon } from '~/ui/components/DocIcons';
 - Designed for speed
 - Support for many image formats (including animated ones)
 - Disk and memory caching
-- Supports [blurhash](https://blurha.sh), a compact representation of a placeholder for an image
+- Supports [blurhash](https://blurha.sh) and [thumbhash](https://evanw.github.io/thumbhash/) - compact representations of a placeholder for an image
```
Comment (Simek) [inline docs/pages/versions/unversioned/sdk/image.mdx:21]:
```suggestion
- Supports [BlurHash](https://blurha.sh) and [ThumbHash](https://evanw.github.io/thumbhash/) - compact representations of a placeholder for an image
```

Names casing nit.

File: packages/expo-image/src/utils/resolveHashString.tsx [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+}
+
+export function resolveBlurhashString(str: string): ImageSource {
```
Comment (Simek) [inline packages/expo-image/src/utils/resolveHashString.tsx:8]: It those methods are exported at package level we need to document them.

**Review by Simek — COMMENTED** [review]
Left few small suggestions related to the docs/content parts.

File: packages/expo-image/src/utils/resolveHashString.tsx (hunk 3 above)
Comment (behenate) [reply packages/expo-image/src/utils/resolveHashString.tsx:5]: Yep, in ThumbHash width and height are automatically encoded into the hash when generating. Aspect ratio (approximate) can be read on the native side.

**Comment by behenate** [comment]
> Also, it's a good practice to include a screenshot in the PR description of the examples added to native-component-list, so the reviewers know how it's going to look like 😉



Good idea! Here is the new example screen:

https://user-images.githubusercontent.com/31368152/229898302-cdcf31c9-c8aa-4a55-b222-8c25a8376609.mp4

**Review by aleqsio — APPROVED** [review]
Looks great!

**Review by lukmccall — APPROVED** [review]
🔥🔥🔥

**Review by tsapeta — APPROVED** [review]
🚀

**Pull request merged**